src/qibo/core/terms.py

```diff
@@ -107,65 +111,72 @@ class SymbolicTerm(HamiltonianTerm):
     """
 
-    def __init__(self, coefficient, factors=[], matrix_map={}):
+    def __init__(self, coefficient, factors=1, symbol_map=None):
         self.coefficient = complex(coefficient)
-        self.factors = factors
-        self.matrix_map = matrix_map
         self._matrix = None
         self._gate = None
         self.hamiltonian = None
-        self.target_qubits = tuple(sorted(self.matrix_map.keys()))
 
-    @classmethod
-    def from_factors(cls, coefficient, factors, symbol_map=None):
-        if factors == 1:
-            return cls(coefficient)
-
-        _factors = []
-        _matrix_map = {}
-        for factor in factors.as_ordered_factors():
-            if isinstance(factor, sympy.Pow):
-                factor, pow = factor.args
-                assert isinstance(pow, sympy.Integer)
-                assert isinstance(factor, sympy.Symbol)
-                pow = int(pow)
-            else:
-                pow = 1
-
-            if symbol_map is not None and factor in symbol_map:
-                from qibo.symbols import Symbol
-                q, matrix = symbol_map.get(factor)
-                factor = Symbol(q, matrix, name=factor.name)
-
-            if isinstance(factor, sympy.Symbol):
-                if isinstance(factor.matrix, K.qnp.tensor_types):
-                    _factors.extend(pow * [factor])
-                    q = factor.target_qubit
-                    if q in _matrix_map:
-                        _matrix_map[q].extend(pow * [factor.matrix])
-                    else:
-                        _matrix_map[q] = pow * [factor.matrix]
+        self.factors = []
+        self.matrix_map = {}
+        if factors != 1:
+            for factor in factors.as_ordered_factors():
+                # check if factor has some power ``power`` so that the corresponding
+                # matrix is multiplied ``pow`` times
+                if isinstance(factor, sympy.Pow):
+                    factor, pow = factor.args
+                    assert isinstance(pow, sympy.Integer)
+                    assert isinstance(factor, sympy.Symbol)
+                    pow = int(pow)
                 else:
-                    coefficient *= factor.matrix
-            elif factor == sympy.I:
-                coefficient *= 1j
-            else: # pragma: no cover
-                raise_error(TypeError, "Cannot parse factor {}.".format(factor))
+                    pow = 1
+
+                # if the user is using ``symbol_map`` instead of qibo symbols,
+                # create the corresponding symbols
+                if symbol_map is not None and factor in symbol_map:
+                    from qibo.symbols import Symbol
+                    q, matrix = symbol_map.get(factor)
+                    factor = Symbol(q, matrix, name=factor.name)
+
+                if isinstance(factor, sympy.Symbol):
+                    if isinstance(factor.matrix, K.qnp.tensor_types):
+                        self.factors.extend(pow * [factor])
```

Comment:
Given that the comments are quite good, please some sentence explaining how and why the pow term is applied to the matrix_map.